typed_python/compiler/type_wrappers/class_wrapper.py

```diff
@@ -850,3 +883,156 @@ def generateConstructor(self, context, out, *args):
+            or self.generate_method_call(context, magic, (l, r)) \
+            or self.convert_comparison(context, l, op, r) \
+            or super().convert_bin_op(context, l, op, r, inplace)
+
+    def convert_comparison(self, context, left, op, right):
+        # TODO: provide nicer translation from op to Py_ comparison codes
+        py_code = 2 if op.matches.Eq else \
+            3 if op.matches.NotEq else \
+            0 if op.matches.Lt else \
+            4 if op.matches.Gt else \
+            1 if op.matches.LtE else \
+            5 if op.matches.GtE else -1
+        if py_code < 0:
+            return None
+        tp_left = context.getTypePointer(left.expr_type.typeRepresentation)
+        tp_right = context.getTypePointer(right.expr_type.typeRepresentation)
+        if tp_left and tp_left == tp_right:
+            if not left.isReference:
+                left = context.push(left.expr_type, lambda x: x.convert_copy_initialize(left))
+            if not right.isReference:
+                right = context.push(right.expr_type, lambda x: x.convert_copy_initialize(right))
+            return context.pushPod(
+                Bool,
+                runtime_functions.class_cmp.call(
+                    tp_left,
+                    left.expr.cast(VoidPtr),
+                    right.expr.cast(VoidPtr),
+                    py_code
+                )
+            )
+        return None
+
+    def convert_bin_op_reverse(self, context, r, op, l, inplace):
+        if op.matches.In:
+            ret = self.generate_method_call(context, "__contains__", (r, l))
+            return (ret and ret.toBool()) \
+                or super().convert_bin_op_reverse(context, r, op, l, inplace)
+
+        magic = "__radd__" if op.matches.Add else \
+            "__rsub__" if op.matches.Sub else \
+            "__rmul__" if op.matches.Mult else \
+            "__rtruediv__" if op.matches.Div else \
+            "__rfloordiv__" if op.matches.FloorDiv else \
+            "__rmod__" if op.matches.Mod else \
+            "__rmatmul__" if op.matches.MatMult else \
+            "__rpow__" if op.matches.Pow else \
+            "__rlshift__" if op.matches.LShift else \
+            "__rrshift__" if op.matches.RShift else \
+            "__ror__" if op.matches.BitOr else \
+            "__rxor__" if op.matches.BitXor else \
+            "__rand__" if op.matches.BitAnd else \
+            ""
+
+        return self.generate_method_call(context, magic, (r, l)) \
+            or super().convert_bin_op_reverse(context, r, op, l, inplace)
+
+    def convert_hash(self, context, expr):
+        y = self.generate_method_call(context, "__hash__", (expr,))
+        if y is not None:
+            return y
+        tp = context.getTypePointer(expr.expr_type.typeRepresentation)
+        if tp:
+            return context.pushPod(Int32, runtime_functions.hash_class.call(expr.nonref_expr.cast(VoidPtr), tp))
```

Comment:
the default implementation in 'HeldClassType' just hashes the class pointer. We should be able to do that directly in compiler code rather than hitting a runtime function for it. See `HeldClass::hash`

Reply:
This has been done.